### ISSUE TITLE
deprecate twisted.internet.gtkreactor

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2833]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2833|
|Type|task|
|Created|2007-09-20 12:07:46Z|
|Branch|https://github.com/twisted/twisted/tree/deprecate-gtkreactor-2833|

twisted.internet.gtkreactor supports integration with gtk 1.x, a GUI toolkit which was last released in 2001 (six years ago at the time of this ticket's creation).  All new applications should be written with gtk 2.x, which is supported by twisted.internet.gtk2reactor.

The specific motivation to deprecate gtkreactor is that it does not provide correct disconnection notification in certain circumstances and it seems unlikely that anyone is going to fix it.

<details><summary>Searchable metadata</summary>

```
trac-id__2833 2833
type__task task
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_deprecate_gtkreactor_2833 branches/deprecate-gtkreactor-2833
branch_author__cyli cyli
status__closed closed
resolution__fixed fixed
component__core core
keywords__easy easy
time__1190290066000000 1190290066000000
changetime__1275245468000000 1275245468000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [21301]) Merge epoll-hup-err-2809-2

Author: exarkun
Reviewer: therve
Fixes [#2809](https://github.com/twisted/twisted/issues/2809)
Refs [[#1780](https://github.com/twisted/twisted/issues/1780)](https://github.com/twisted/twisted/issues/1780)
Refs [#11168](https://github.com/twisted/twisted/issues/11168)
Refs [[#2833](https://github.com/twisted/twisted/issues/2833)](https://github.com/twisted/twisted/issues/2833)

Change the error (EPOLLHUP, EPOLLERR) checking code so that it can
handle the case where they are both set, rather than only checking
the cases where one or the other is set.  This allows epollreactor
to properly detect and report connections which are lost uncleanly
and also are not being polled for input events.

The test case for this is slightly convoluted for a number of
reasons:

  * gtkreactor cannot pass the new test, and I am not fixing that
    reactor in this branch.  The test will skip when that reactor
    is in use.  #2833 was created as a result of this in order to
    be able to remove the broken gtkreactor eventually.
  * If a file descriptor is being polled for neither read nor write
    events, it will not be polled for hang-up events either.  This
    is a known issue, [#11168](https://github.com/twisted/twisted/issues/11168).  When it is fixed, the test will need
    to be updated.
  * In order to exercise the case where the event reported from epoll
    is only HUP|ERR (specifically, not IN), the test pauses one of
    the involved transports.  However, due to #1780, it must use a
    timing hack to accomplish this.  Once that issue is resolved,
    it should be possible to remove this hack.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_2"></a>|@jyknight commented|
|-|-|

Depends on [#3340](https://github.com/twisted/twisted/issues/3340).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_4"></a>|@cyli commented|
|-|-|

(In [28990]) Branching to 'deprecate-gtkreactor-2833'

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_5"></a>|@cyli commented|
|-|-|

(In [28993]) Deprecated all the attributes in twisted.internet.gtkreactor and added a test
that tests that all attributes are indeed deprecated.

refs [#2833](https://github.com/twisted/twisted/issues/2833)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_7"></a>|@cyli commented|
|-|-|

(In [28994]) Added topfile for this deprecation.

refs [#2833](https://github.com/twisted/twisted/issues/2833)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @cyli|
|-|-|

Yay.  This is exciting.

I expect it's also going to be a bit painful, though. :/

For example, when I run the full test suite now, all of the tests in `twisted.internet.test.test_gtkreactor` fail like this:

```
[ERROR]: twisted.internet.test.test_gtkreactor.GtkReactorDeprecation.test_portableInstall

Traceback (most recent call last):
  File "/home/exarkun/Projects/Twisted/trunk/twisted/internet/test/test_gtkreactor.py", line 93, in test_portableInstall
    from twisted.internet import gtkreactor
  File "/home/exarkun/Projects/Twisted/trunk/twisted/internet/gtkreactor.py", line 21, in <module>
    pygtk.require('1.2')
  File "/usr/lib/pymodules/python2.6/pygtk.py", line 81, in require
    "pygtk.require() must be called before importing gtk"
exceptions.AssertionError: pygtk.require() must be called before importing gtk
```

I thought this was because pygtk2 had already been loaded (for gtk2reactor, presumably).  But when I run just `twisted.internet.test.test_gtkreactor`, I get the same thing.  Perhaps that's because I don't have gtk 1 installed at all?  But then I'd expect to see an error message saying that, so I'm a bit confused.

So:

  1. Can these tests play nicely with the gtk2reactor tests?  Do we need to run them in a child process instead?  Or (gasp) just give automated testing a pass in this case?
  1. Why don't the tests fail by saying gtk 1 isn't installed, instead of this other confusing message?
  1. The tests need to skip if gtk 1 really isn't installed.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_9"></a>|@cyli commented|
|-|-|

Hmm... these tests passed on my mac, which has neither gtk nor pygtk installed.  I assume because `gtkreactor` doesn't complain if pygtk isn't installed at all (it seems to pass if there is an ImportError or AttributeError, but not an AssertionError).

I was told that it was very rare to have gtk 1 installed these days, and hence the deprecation would be difficult to test if I do not simply simulate having a gtk1 module.

So: 

1) It seems to be the import assertion in gtkreactor that's the problem, so importing gtkreactor while pygtk!=1.2 is installed will cause a problem.

2 & 3) I was expecting that on most environments these tests are run on,  pygtk 1.2 and gtk1 would not be installed.  I was hoping to ensure that the deprecation tests happen anyway.

I am thinking of 3 possible solutions:

a) stub out the pygtk import in the test setup as I do with the gtk import

b) skip the tests entirely if gtk1 is not installed (and remove my gtk import stub)

c) leave the stub but skip the test if gtk2 is installed.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_10"></a>|@cyli commented|
|-|-|

(In [29061]) Stubbing out pygtk as well as gtk

refs [#2833](https://github.com/twisted/twisted/issues/2833)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_11"></a>|@cyli commented|
|-|-|

(In [29062]) Simplify replacing sys.modules code in the test setUp/tearDown methods.

refs [#2833](https://github.com/twisted/twisted/issues/2833)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_12"></a>|@cyli removed owner|
|-|-|

Ran the new gtkreactor tests (`twisted.internet.test.test_gtkreactor`) on a machine with gtk2.0 installed and the tests passed.  Also ran the full twisted suite on the same machine and the tests passed.

Went with stubbing out the modules, because otherwise the deprecation code would probably never get tested (is there a gtk1 build slave?).  All I did was add deprecation messages - and tests for those deprecation messages.  There are probably tests for gtkreactor functionality elsewhere and those are probably skipped, but I'm not sure.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve set owner to @cyli|
|-|-|

 * Both the version object and the message should be created at the module level, and shared between call to `deprecate.deprecatedModuleAttribute` to remove duplication.

Please merge once fixed!

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_14"></a>|@cyli commented|
|-|-|

(In [29097]) Version object and message created at the module level to be shared between calls to deprecate.deprecatedModuleAttribute

refs [#2833](https://github.com/twisted/twisted/issues/2833)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_15"></a>|@cyli set status to `closed`|
|-|-|

(In [29100]) Merge deprecate-gtkreactor-2833: deprecate twisted.internet.gtkreactor 

Author: cyli
Reviewer: exarkun, therve
Fixes: [#2833](https://github.com/twisted/twisted/issues/2833)

Deprecated all the attributes in twisted.internet.gtkreactor and added a test to test that all attributes are indeed deprecated.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
